File: packages/appwsgi/appwsgi%201057.tar.bz2/appwsgi/www/order/order.py
# Copyright(c) gert.cuykens@gmail.com
from frame import Frame
from session import Session
from db import Db
from xml.dom import minidom

def application(environ, response):
    cookie = environ.get('HTTP_COOKIE','')
    document = environ['wsgi.input'].read()
    db = Db()
    v = Frame.read(document)
    s = Session(db,cookie,v['gid'])

    if s.GID=='guest':
        if   v['cmd']=='list'  : db.execute("SELECT uid,oid,bid,time FROM shop_orders WHERE uid=?",(v['uid'],))
        elif v['cmd']=='name'  : db.execute("SELECT uid,name,adress,city,country,phone FROM users WHERE uid=?",(v['uid'],))
        elif v['cmd']=='cart'  : db.execute("SELECT products FROM shop_orders WHERE oid=? AND uid=?",(v['oid'],v['uid'],))
        elif v['cmd']=='find'  : db.execute("SELECT * FROM shop_products WHERE description LIKE ?",('%'+v['pna']+'%',))
        elif v['cmd']=='pid'   : db.execute("SELECT * FROM shop_products WHERE pid=?",(v['pid'],))
        elif v['cmd']=='status': db.execute("SELECT * FROM shop_status")
        elif v['cmd']=='stats' : db.execute("SELECT * FROM shop_orders WHERE bid=? AND uid=?",(v['bid'],v['uid']))

        elif v['cmd']=='insert': 
             db.execute("INSERT INTO shop_orders (products,uid,bid,time) VALUES (?,?,?,DATETIME('NOW'))",(v['rec'],v['uid'],3))
             oid=db.LASTROWID
             insert(db,v['rec'],oid)
             db.LASTROWID=oid

        #elif v['cmd']=='update':
             #update(db,v['rec'],v['oid']) 
             #db.execute("UPDATE shop_orders SET products=? WHERE oid=? AND uid=?",(v['rec'],v['oid'],v['uid']))
             #restock(db)

        #elif v['cmd']=='remove': 
             #remove(db,v['oid'])
             #db.execute("DELETE FROM shop_orders WHERE oid=? AND uid=?",(v['oid'],v['uid']))
             #restock(db)

    if s.GID=='admin':
        if   v['cmd']=='list'  : db.execute("SELECT uid,oid,bid,time FROM shop_orders WHERE uid=?",(v['uida'],))
        elif v['cmd']=='name'  : db.execute("SELECT uid,name,adress,city,country,phone FROM users WHERE name LIKE ?",('%'+v['una']+'%',))
        elif v['cmd']=='cart'  : db.execute("SELECT products FROM shop_orders WHERE oid=?",(v['oid'],))
        elif v['cmd']=='find'  : db.execute("SELECT * FROM shop_products WHERE description LIKE ?",('%'+v['pna']+'%',))
        elif v['cmd']=='pid'   : db.execute("SELECT * FROM shop_products WHERE pid=?",(v['pid'],))
        elif v['cmd']=='status': db.execute("SELECT * FROM shop_status")
        elif v['cmd']=='stats' : db.execute("SELECT * FROM shop_orders WHERE bid=?",(v['bid'],))   
        
        elif v['cmd']=='insert':
             db.execute("INSERT INTO shop_orders (products,uid,bid,time) VALUES (?,?,?,DATETIME('NOW'))",(v['rec'],v['uida'],3))
             oid=db.LASTROWID
             insert(db,v['rec'],oid)
             db.LASTROWID=oid

        elif v['cmd']=='update':
             update(db,v['rec'],v['oid'])
             db.execute("UPDATE shop_orders SET products=?, uid=? WHERE oid=?",(v['rec'],v['uida'],v['oid']))
             restock(db)

        elif v['cmd']=='remove':
             remove(db,v['oid'])
             db.execute("DELETE FROM shop_orders WHERE oid=?",(v['oid'],))
             restock(db)

    xml = '<?xml version="1.0" encoding="UTF-8"?>\n'
    xml+= '<root>\n'
    xml+= ' <cmd>'+str(v['cmd'])+'</cmd>\n'
    xml+= ' <gid>'+str(s.GID)+'</gid>\n'
    xml+= ' <uid>'+str(s.UID)+'</uid>\n'
    xml+= ' <sid>'+str(s.SID)+'</sid>\n'
    xml+= ' <exp>'+str(s.EXP)+'</exp>\n'
    xml+= Frame.write(db)
    xml+= '</root>'

    response('200 OK', [('Content-type', 'text/xml'), ('Set-Cookie', s.COOKIE)])
    return [xml]

def insert(db,doc,oid):
    document = minidom.parseString(doc)
    status=3
    for i,pid in enumerate(document.getElementsByTagName('pid')):
        p=pid.childNodes[0].nodeValue
        q=document.getElementsByTagName('qty')[i].childNodes[0].nodeValue
        db.execute("UPDATE shop_products SET qty=qty-? WHERE pid=?",(q,p))
        db.execute("SELECT qty FROM shop_products WHERE pid=?",(p,))
        stock=int(db.fetch()[0][0])
        if stock < 0: status=2
    db.execute("UPDATE shop_orders SET bid=? WHERE oid=?",(status,oid))

def update(db,doc,oid):
    db.execute("SELECT products FROM shop_orders WHERE oid=?",(oid,))
    document = minidom.parseString(db.fetch()[0][0])
    documenn = minidom.parseString(doc)
    status=3
    for j,pn in enumerate(documenn.getElementsByTagName('pid')):
        p=pn.childNodes[0].nodeValue
        q=documenn.getElementsByTagName('qty')[j].childNodes[0].nodeValue
        u=bool(1)
        for i,pid in enumerate(document.getElementsByTagName('pid')):
            if p == pid.childNodes[0].nodeValue: u = bool(0)
        if u : db.execute("UPDATE shop_products SET qty=qty-? WHERE pid=?",(q,p))
        db.execute("SELECT qty FROM shop_products WHERE pid=?",(p,))
        stock=int(db.fetch()[0][0])
        if stock < 0: status=2
    for i,pid in enumerate(document.getElementsByTagName('pid')):
        p=pid.childNodes[0].nodeValue
        q=document.getElementsByTagName('qty')[i].childNodes[0].nodeValue
        u=0
        for j,pn in enumerate(documenn.getElementsByTagName('pid')):
            if  p == pn.childNodes[0].nodeValue:
                u=documenn.getElementsByTagName('qty')[j].childNodes[0].nodeValue
        q=int(q)-int(u)
        if q != 0: db.execute("UPDATE shop_products SET qty=qty+? WHERE pid=?",(q,p))
        db.execute("SELECT qty FROM shop_products WHERE pid=?",(p,))
        stock=int(db.fetch()[0][0])
        if stock < 0: status=2
    db.execute("UPDATE shop_orders SET bid=? WHERE oid=?",(status,oid))

def remove(db,oid):
    db.execute("SELECT products FROM shop_orders WHERE oid=?",(oid,))
    document = minidom.parseString(db.fetch()[0][0])
    for i,pid in enumerate(document.getElementsByTagName('pid')):
        p=pid.childNodes[0].nodeValue
        q=document.getElementsByTagName('qty')[i].childNodes[0].nodeValue
        db.execute("UPDATE shop_products SET qty=qty+? WHERE pid=?",(q,p))
    db.execute("UPDATE shop_orders SET bid=1 WHERE oid=?",(oid,))

def restock(db):    
    db.execute("SELECT products,oid FROM shop_orders WHERE bid=2 OR bid=3 ORDER BY time ASC",())
    f = db.fetch()
    for r in f:
        doc,oid=r
        document = minidom.parseString(doc)
        status = 3
        for i,pid in enumerate(document.getElementsByTagName('pid')):
            p=pid.childNodes[0].nodeValue
            db.execute("SELECT qty FROM shop_products WHERE pid=?",(p,))
            try: stock=int(db.fetch()[0][0])
            except: stock=0
            if stock < 0: status=2
        db.execute("UPDATE shop_orders SET bid=? WHERE oid=?",(status,oid))

File: packages/appwsgi/appwsgi%201317.tar.bz2/appwsgi/www/order/order.js
// Copyright(c) gert.cuykens@gmail.com
products=
{
 'id':'products',
 'rec':[],
 'des':[],
 'onselect':function(){}
}

order=
{
 'uid':'',
 'list':function(v)
 {
  document.getElementById('products-dg').innerHTML=''
  products.rec=v['rec']
  products.des=v['des']
  products.onselect=function()
  {
   var gid='guest'; try{gid=document.URL.match(/-(.*)/)[1]} catch(e){}
   location.href='../invoice/invoice.htm?'+products.selection[1]+'-'+gid
  }
  dg.init(products)
 },

 'stat':function(v)
 {
  document.getElementById('products-dg').innerHTML=''
  products.rec=v['rec']
  products.des=v['des']
  products.onselect=function()
  {
   var gid='guest'; try{gid=document.URL.match(/-(.*)/)[1]} catch(e){}
   http.fml =function(v)
   {
    document.getElementById('products-dg').innerHTML=''
    products.rec=v['rec']
    products.des=v['des']
    products.onselect=function()
    {
     var gid='guest'; try{gid=document.URL.match(/-(.*)/)[1]} catch(e){}
     location.href='../invoice/invoice.htm?'+products.selection[1]+'-'+gid
    }
    dg.init(products)
   }
   http.xml ='{"cmd":"stats",\n'
   http.xml+=' "gid":"'+gid+'",\n'
   http.xml+=' "bid":"'+products.selection[0]+'"}'
   http.url ='order.wsgi'
   http.req ='POST'
   http.send()
  }
  dg.init(products)
 },

 'name':function(v)
 {
  document.getElementById('products-dg').innerHTML=''
  products.rec=v['rec']
  products.des=v['des']
  products.onselect=function()
  {
   order.uid=products.selection[0]
   var gid='guest'; try{gid=document.URL.match(/-(.*)/)[1]} catch(e){}
   http.fml = function(v)
   {
    document.getElementById('products-dg').innerHTML=''
    products.rec=v['rec']
    products.des=v['des']
    products.onselect=function()
    {
     var gid='guest'; try{gid=document.URL.match(/-(.*)/)[1]} catch(e){}
     location.href='../invoice/invoice.htm?'+products.selection[1]+'-'+gid
    }
    dg.init(products)
   }
   http.xml ='{"cmd":"list",\n'
   http.xml+=' "gid":"'+gid+'",\n'
   http.xml+=' "uid":"'+order.uid+'"}'
   http.url ='order.wsgi'
   http.req ='POST'
   http.send()
  }
  dg.init(products)
 },

 'menu':function()
 {
  var f1=document.createElement('input')
  f1.id='f1'
  f1.type='button'
  f1.value='status'
  f1.onclick=function()
  {
   var gid='guest'; try{gid=document.URL.match(/-(.*)/)[1]} catch(e){}
   http.fml =order.stat
   http.xml ='{"cmd":"status",\n'
   http.xml+=' "gid":"'+gid+'"}'
   http.url ='order.wsgi'
   http.req ='POST'
   http.send()
   if(gid=='admin')document.getElementById('f1').focus()
  }

  var f2=document.createElement('input')
  f2.id='f2'
  f2.type='button'
  f2.value='shop'
  f2.onclick=function()
  {
   var gid='guest'; try{gid=document.URL.match(/-(.*)/)[1]} catch(e){}
   location.href='../shop/shop.htm?'+order.uid+'-'+gid
  }

  var f3=document.createElement('input')
  f3.id='session'
  f3.type='button'
  f3.value='login'

  var menu=document.getElementById('menu')
  menu.id='menu'
  menu.appendChild(f1)
  menu.appendChild(f2)
  menu.appendChild(f3)
 },

 'select':function()
 {
  var gid='guest'; try{gid=document.URL.match(/-(.*)/)[1]} catch(e){}
  var una=''; try{una=document.URL.match(/\?(.*)-/)[1]} catch(e){}
  http.fml =order.name
  http.xml ='{"cmd":"name",\n'
  http.xml+=' "gid":"'+gid+'",\n'
  http.xml+=' "una":"'+una+'"}'
  http.url ='order.wsgi'
  http.req ='POST'
  if(gid=='admin'){http.send();return 0}
  http.fml = function(v)
  {
   document.getElementById('products-dg').innerHTML=''
   products.rec=v['rec']
   products.des=v['des']
   products.onselect=function()
   {
    var gid='guest'; try{gid=document.URL.match(/-(.*)/)[1]} catch(e){}
    location.href='../invoice/invoice.htm?'+products.selection[1]+'-'+gid
   }
   dg.init(products)
  }
  http.xml ='{"cmd":"list",\n'
  http.xml+=' "gid":"guest"}'
  http.url ='order.wsgi'
  http.req ='POST'
  http.send()
 }
}
